libs/pilight/hardware/mqtt.c

```diff
@@ -0,0 +1,1358 @@
+					protocols = protocols->next;
+				}
+
+			} else {
+				logprintf(LOG_DEBUG, "mqtt_listener: device \"%s\" doesn't exist in config", array[n - 1]);
+			}
+
+		} break;
+
+		case -3: { //hardware module not ready
+			logprintf(LOG_NOTICE, "mqtt_listener: client \"%s\" tried to register, but it seems that mqtt module was not configured", LISTENER_CLIENTID);
+			exit(EXIT_FAILURE);
+			uv_timer_start(listener_timer_req, (void (*)(uv_timer_t *))listener_wait, MIN_DELAY, 0);
+		} break;
+
+		case -2: { // Connection lost
+			logprintf(LOG_NOTICE, "mqtt_listener: connection for \"%s\"lost, will try to register again in %d seconds", LISTENER_CLIENTID, listener_delay / 1000);
+			uv_timer_start(listener_timer_req, (void (*)(uv_timer_t *))listener_wait, listener_delay, 0);
+			listener_delay *= 2;
+			if(listener_delay > MAX_DELAY) {
+				listener_delay = MAX_DELAY;
+			}
+		}
+	}
+}
+
+static unsigned short int mqttHwInit(void) {
+	initialized = 1;
+	if(listener_topic != NULL && strlen(listener_topic) > 0) {
+		if((listener_timer_req = MALLOC(sizeof(uv_timer_t))) == NULL) {
+			OUT_OF_MEMORY /*LCOV_EXCL_LINE*/
+		}
+		listener_timer_req->data = NULL; //no data needed
+		uv_timer_init(uv_default_loop(), listener_timer_req);
+
+		logprintf(LOG_DEBUG, "mqtt_listener: registering clientid \"%s\" for receive", LISTENER_CLIENTID);
+		mqtt_sub(LISTENER_CLIENTID, default_host, default_port, default_user, default_pass, listener_topic, 1, 0, listener_callback, NULL); //no need for userdata, because this client is part of the main module
+	}
+
+	return EXIT_SUCCESS;
+}
+
+static unsigned short mqttSettings(struct JsonNode *json) {
+
+	if(strcmp(json->key, "host") == 0) {
+		default_host = NULL;
+		if(json->tag == JSON_STRING) {
+			if((default_host = MALLOC(strlen(json->string_) + 1)) == NULL) {
+				OUT_OF_MEMORY /*LCOV_EXCL_LINE*/
+			}
+			strcpy(default_host, json->string_);
+		} else {
+			return EXIT_FAILURE;
+		}
+	}
+
+	if(strcmp(json->key, "user") == 0) {
+		default_user = NULL;
+		if(json->tag == JSON_STRING) {
+			if((default_user = MALLOC(strlen(json->string_) + 1)) == NULL) {
+				OUT_OF_MEMORY /*LCOV_EXCL_LINE*/
+			}
+			strcpy(default_user, json->string_);
+		} else {
+			return EXIT_FAILURE;
+		}
+	}
+
+	if(strcmp(json->key, "pass") == 0) {
+		default_pass = NULL;
+		if(json->tag == JSON_STRING) {
+			if((default_pass = MALLOC(strlen(json->string_) + 1)) == NULL) {
+				OUT_OF_MEMORY /*LCOV_EXCL_LINE*/
+			}
+			strcpy(default_pass, json->string_);
+		} else {
+			return EXIT_FAILURE;
+		}
+	}
+
+	if(strcmp(json->key, "port") == 0) {
+		if(json->tag == JSON_NUMBER) {
+			default_port = (int)round(json->number_);
+		} else {
+			return EXIT_FAILURE;
+		}
+	}
+
+	if(strcmp(json->key, "listener_topic") == 0) {
+		if(json->tag == JSON_STRING) {
+			if((listener_topic = MALLOC(strlen(json->string_) + 1)) == NULL) {
+				OUT_OF_MEMORY /*LCOV_EXCL_LINE*/
+			}
+			strcpy(listener_topic, json->string_);
+		} else {
+			return EXIT_FAILURE;
+		}
+	}
+
```

Comment:
Each code block can return in an `EXIT_FAILURE`  but the memory already allocated isn't freed.